src/Features/Core/Portable/CodeFixesAndRefactorings/CodeActionRequestPriorityProvider.cs

```diff
@@ -0,0 +1,146 @@
+            // by high pri or normal pri fixers.
+            if (analyzer.IsCompilerAnalyzer())
+                return true;
+
+            // Check if we are computing diagnostics for 'CodeActionRequestPriority.Low' and
+            // this analyzer was de-prioritized to low priority bucket.
+            if (Priority == CodeActionRequestPriority.Low &&
+                _lowPriorityAnalyzers.Contains(analyzer))
+            {
+                return true;
+            }
+
+            var analyzerPriority = analyzer is IBuiltInAnalyzer { RequestPriority: var requestPriority }
+                ? requestPriority
+                : CodeActionRequestPriority.Normal;
+
+            return Priority == analyzerPriority;
+        }
+
+        /// <summary>
+        /// Returns true if this is an analyzer that is a candidate to be de-prioritized to
+        /// <see cref="CodeActionRequestPriority.Low"/> <see cref="Priority"/> for improvement in analyzer
+        /// execution performance for priority buckets above 'Low' priority.
+        /// </summary>
+        /// <remarks>
+        /// Based on performance measurements, currently only analyzers which register SymbolStart/End actions
+        /// or SemanticModel actions are considered candidates to be de-prioritized. However, these semantics
+        /// could be changed in future based on performance measurements.
+        /// </remarks>
+        public static async Task<bool> IsCandidateForDeprioritizationBasedOnRegisteredActionsAsync(
+            DiagnosticAnalyzer analyzer,
+            CompilationWithAnalyzers? compilationWithAnalyzers,
+            CancellationToken cancellationToken)
+        {
+            // We deprioritize SymbolStart/End and SemanticModel analyzers from 'Normal' to 'Low' priority bucket,
+            // as these are computationally more expensive.
+            // Note that we never de-prioritize compiler analyzer, even though it registers a SemanticModel action.
+            if (compilationWithAnalyzers != null && !analyzer.IsWorkspaceDiagnosticAnalyzer())
+            {
+                var telemetryInfo = await compilationWithAnalyzers.GetAnalyzerTelemetryInfoAsync(analyzer, cancellationToken).ConfigureAwait(false);
+                return (telemetryInfo?.SymbolStartActionsCount > 0 || telemetryInfo?.SemanticModelActionsCount > 0) &&
+                    !analyzer.IsCompilerAnalyzer();
+            }
+
+            return false;
+        }
+
+        /// <summary>
+        /// Tracks the given <paramref name="analyzer"/> as a de-prioritized analyzer that should be moved to
+        /// <see cref="CodeActionRequestPriority.Low"/> bucket.
+        /// </summary>
+        public void TrackDeprioritizedAnalyzer(DiagnosticAnalyzer analyzer)
+        {
+            Debug.Assert(Priority == CodeActionRequestPriority.Normal);
+            _lowPriorityAnalyzers.Add(analyzer);
+        }
```

Comment:
i hate this, especially given that we have a static readonly provider.  Either no static-readonly (everyone gets a fresh instance), or make these immtuable types and have this return a fresh forked instance.